File: Hash Table/Hash_table.py
class HashTable:
    def __init__(self, size=5):
        self.size = size
        self.table = [[] for _ in range(size)]  # List of lists for chaining
    
    def _hash_function(self, key):
        """ Simple hash function: key mod table size """
        return hash(key) % self.size
    
    def insert(self, key, value):
        """ Inserts a key-value pair into the hash table """
        index = self._hash_function(key)
        # Check if the key already exists, if yes, update the value
        print(f"Added key:{key} to the HashTable")
        for pair in self.table[index]:
            if pair[0] == key:
                pair[1] = value
                return
        # Otherwise, append the new key-value pair used chaining i.e separate chaining -> linked list method to avoid collision.
        self.table[index].append([key, value])
    
    def get(self, key):
        """ Retrieves the value associated with the key """
        index = self._hash_function(key)
        for pair in self.table[index]:
            if pair[0] == key:
                return pair[1]
        return None  # Return None if the key is not found
    
    def display(self):
        """ Displays the hash table contents """
        for i, bucket in enumerate(self.table):
            print(f"Index {i}: {bucket}")

hash_table = HashTable()

hash_table.insert("Name", "Bhuvam")
hash_table.insert("Age", 20)
hash_table.insert("Degree", "B.Tech")
hash_table.insert("Specialization", "IT")
hash_table.insert("City", "CBE")

# Retrieve values
print("Specialization:", hash_table.get("Specialization")) 
print("City:", hash_table.get("city")) # keys are case-sensitive "City" differs from "city".

# Display hash table
hash_table.display()

File: Hash Table/Hash_Table.java
import java.util.Arrays;

class Main {
    private int size = 10;
    private String[] keys;
    private String[] values;
    
    public Main() {
        keys = new String[size];
        values = new String[size];
    }
    
    // Hash function
    private int hashFunction(String key) {
        return key.hashCode() % size;
    }
    
    // Insert key-value pair into the hash table
    public void insert(String key, String value) { // key = "name", value = "Alice"
        int index = hashFunction(key);// index = 0
        while (keys[index] != null) {
            if (keys[index].equals(key)) {
                values[index] = value;  // Update value if key exists
                return;
            }
            index = (index + 1) % size;  // Linear probing
        }
        keys[index] = key;
        values[index] = value;
    }
    
    // Get the value associated with the key
    public String get(String key) {
        int index = hashFunction(key);
        while (keys[index] != null) {
            if (keys[index].equals(key)) {
                return values[index];
            }
            index = (index + 1) % size;  // Linear probing
        }
        return null;  // Key not found
    }
    
    // Display hash table contents
    public void display() {
        System.out.println("Hash Table:");
        for (int i = 0; i < size; i++) {
            if (keys[i] != null) {
                System.out.println("Index " + i + ": Key = " + keys[i] + ", Value = " + values[i]);
            } else {
                System.out.println("Index " + i + ": Empty");
            }
        }
    }
    
    public static void main(String[] args) {
        Main hashTable = new Main();
        
        // Insert key-value pairs
        hashTable.insert("name", "Alice");
        hashTable.insert("name", "Bharath");
        hashTable.insert("age", "30");
        hashTable.insert("city", "New York");
        
        // Retrieve values
        System.out.println("Name: " + hashTable.get("name"));  // Output: Bharath
        System.out.println("Age: " + hashTable.get("age"));    // Output: 30
        System.out.println("City: " + hashTable.get("city"));  // Output: New York
        
        // Display the hash table
        hashTable.display();
    }
}
